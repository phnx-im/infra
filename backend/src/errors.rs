// SPDX-FileCopyrightText: 2024 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use thiserror::Error;
use tracing::error;

#[derive(Debug, Error)]
pub enum StorageError {
    #[error(transparent)]
    Database(#[from] DatabaseError),
    #[error("Error deserializing column: {0}")]
    Serde(#[from] phnxtypes::codec::Error),
}

impl From<sqlx::Error> for StorageError {
    fn from(e: sqlx::Error) -> Self {
        Self::Database(e.into())
    }
}

impl From<Box<dyn std::error::Error + Send + Sync>> for StorageError {
    fn from(e: Box<dyn std::error::Error + Send + Sync>) -> Self {
        Self::Database(e.into())
    }
}

impl From<StorageError> for tonic::Status {
    fn from(e: StorageError) -> Self {
        match e {
            StorageError::Database(error) => {
                error!(%error, "Database error");
                Self::internal("Database error")
            }
            StorageError::Serde(error) => {
                error!(%error, "Seriazation error");
                Self::internal("Seriazation error")
            }
        }
    }
}

#[derive(Debug, Error)]
pub enum DatabaseError {
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Dynamic(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// General error while accessing the requested queue.
#[derive(Error, Debug)]
pub(super) enum QueueError {
    #[error(transparent)]
    Storage(#[from] StorageError),
    /// Mismatching sequence numbers.
    #[error("Mismatching sequence numbers.")]
    SequenceNumberMismatch,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
}

impl From<sqlx::Error> for QueueError {
    fn from(e: sqlx::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl From<phnxtypes::codec::Error> for QueueError {
    fn from(e: phnxtypes::codec::Error) -> Self {
        Self::Storage(e.into())
    }
}
