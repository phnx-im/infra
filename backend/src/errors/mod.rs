// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use displaydoc::Display;
use mls_assist::{
    group, memory_provider::MlsAssistMemoryStorage, openmls::group::MergeCommitError,
};
use thiserror::Error;
use tokio::sync::mpsc;
use tonic::Status;
use tracing::error;

use aircommon::codec::PersistenceCodec;

pub(crate) mod auth_service;
pub(crate) mod qs;

pub(crate) type CborMlsAssistStorage = MlsAssistMemoryStorage<PersistenceCodec>;

#[derive(Debug, Error)]
pub enum StorageError {
    #[error(transparent)]
    Database(#[from] DatabaseError),
    #[error("Error deserializing column: {0}")]
    Serde(#[from] aircommon::codec::Error),
}

impl From<sqlx::Error> for StorageError {
    fn from(e: sqlx::Error) -> Self {
        Self::Database(e.into())
    }
}

impl From<Box<dyn std::error::Error + Send + Sync>> for StorageError {
    fn from(e: Box<dyn std::error::Error + Send + Sync>) -> Self {
        Self::Database(e.into())
    }
}

impl From<StorageError> for Status {
    fn from(error: StorageError) -> Self {
        error!(%error, "storage error");
        match error {
            StorageError::Database(_) => Self::internal("Database error"),
            StorageError::Serde(_) => Self::internal("Seriazation error"),
        }
    }
}
#[derive(Debug, Error)]
pub enum DatabaseError {
    #[error(transparent)]
    Sqlx(#[from] sqlx::Error),
    #[error(transparent)]
    Dynamic(#[from] Box<dyn std::error::Error + Send + Sync>),
}

/// General error while accessing the requested queue.
#[derive(Error, Debug, Display)]
pub(super) enum QueueError {
    /// {0:?}
    Storage(#[from] StorageError),
    /// Mismatching sequence numbers
    SequenceNumberMismatch,
    /// Payload receiver closed
    PayloadReceiverClosed,
}

impl From<sqlx::Error> for QueueError {
    fn from(e: sqlx::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl From<aircommon::codec::Error> for QueueError {
    fn from(e: aircommon::codec::Error) -> Self {
        Self::Storage(e.into())
    }
}

impl<T> From<mpsc::error::SendError<T>> for QueueError {
    fn from(_: mpsc::error::SendError<T>) -> Self {
        Self::PayloadReceiverClosed
    }
}

impl From<QueueError> for Status {
    fn from(error: QueueError) -> Self {
        let msg = error.to_string();
        match error {
            QueueError::Storage(error) => {
                error!(%error, "storage error");
                Self::internal(msg)
            }
            QueueError::SequenceNumberMismatch => Self::internal(msg),
            QueueError::PayloadReceiverClosed => Self::internal(msg),
        }
    }
}

/// Potential errors when performing a group operation.
#[derive(Debug, Error)]
pub(crate) enum GroupOperationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package")]
    MissingQueueConfig,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupOperationError> for Status {
    fn from(e: GroupOperationError) -> Self {
        let msg = e.to_string();
        match e {
            GroupOperationError::LibraryError
            | GroupOperationError::ProcessingError
            | GroupOperationError::InvalidMessage => Status::internal(msg),
            GroupOperationError::MissingQueueConfig | GroupOperationError::IncompleteWelcome => {
                Status::invalid_argument(msg)
            }
            GroupOperationError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when joining a connection group.
#[derive(Debug, Error)]
pub(crate) enum JoinConnectionGroupError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    /// Not a connection group.
    #[error("Not a connection group")]
    NotAConnectionGroup,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<JoinConnectionGroupError> for Status {
    fn from(e: JoinConnectionGroupError) -> Self {
        let msg = e.to_string();
        match e {
            JoinConnectionGroupError::InvalidMessage
            | JoinConnectionGroupError::NotAConnectionGroup => Status::invalid_argument(msg),
            JoinConnectionGroupError::ProcessingError => Status::internal(msg),
            JoinConnectionGroupError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when deleting a group.
#[derive(Debug, Error)]
pub(crate) enum GroupDeletionError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<GroupDeletionError> for Status {
    fn from(e: GroupDeletionError) -> Self {
        let msg = e.to_string();
        match e {
            GroupDeletionError::InvalidMessage => Status::invalid_argument(msg),
            GroupDeletionError::ProcessingError => Status::internal(msg),
            GroupDeletionError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when processing a self remove proposal.
#[derive(Debug, Error)]
pub(crate) enum ClientSelfRemovalError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ClientSelfRemovalError> for Status {
    fn from(e: ClientSelfRemovalError) -> Self {
        let msg = e.to_string();
        match e {
            ClientSelfRemovalError::InvalidMessage => Status::invalid_argument(msg),
            ClientSelfRemovalError::ProcessingError => Status::internal(msg),
            ClientSelfRemovalError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}

/// Potential errors when resyncing a client.
#[derive(Debug, Error)]
pub(crate) enum ResyncClientError {
    /// Invalid assisted message.
    #[error("Invalid assisted message")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message")]
    ProcessingError,
    #[error("Error merging commit")]
    MergeCommitError(#[from] MergeCommitError<group::errors::StorageError<CborMlsAssistStorage>>),
}

impl From<ResyncClientError> for Status {
    fn from(e: ResyncClientError) -> Self {
        let msg = e.to_string();
        match e {
            ResyncClientError::InvalidMessage => Status::invalid_argument(msg),
            ResyncClientError::ProcessingError => Status::internal(msg),
            ResyncClientError::MergeCommitError(merge_commit_error) => {
                error!(%merge_commit_error, "failed merging commit");
                Status::internal(msg)
            }
        }
    }
}
