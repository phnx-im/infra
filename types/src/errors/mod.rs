// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use mls_assist::{
    group::errors::StorageError, memory_provider::MlsAssistMemoryStorage,
    openmls::group::MergeCommitError,
};
use thiserror::Error;
use tracing::error;
use version::VersionError;

use crate::codec::PhnxCodec;

pub mod auth_service;
pub mod qs;
pub mod version;

pub type CborMlsAssistStorage = MlsAssistMemoryStorage<PhnxCodec>;

/// Potential errors when performing a group operation.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum GroupOperationError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package.")]
    MissingQueueConfig,
    /// Failed to retrieve QS verifying key.
    #[error("Failed to retrieve QS verifying key.")]
    FailedToObtainVerifyingKey,
    /// User added twice.
    #[error("User added twice.")]
    DuplicatedUserAddition,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when updating a client.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientUpdateError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when processing a message.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum DsProcessingError {
    /// API Version error
    #[error(transparent)]
    Api(#[from] VersionError),
    /// Failed to distribute message to other members
    #[error("Failed to distribute message to other members")]
    DistributionError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Invalid signature.
    #[error("Invalid signature.")]
    InvalidSignature,
    /// Group not found.
    #[error("Group not found.")]
    GroupNotFound,
    /// Could not decrypt group state.
    #[error("Could not decrypt group state.")]
    CouldNotDecrypt,
    /// Could not encrypt group state.
    #[error("Could not decrypt group state.")]
    CouldNotEncrypt,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    /// Invalid sender type.
    #[error("Invalid sender type")]
    InvalidSenderType,
    /// Error storing encrypted group state.
    #[error("Error storing encrypted group state.")]
    StorageError,
    /// Error creating group.
    #[error("Failed to create group: Group ID not reserved")]
    UnreservedGroupId,
    /// Error updating client.
    #[error(transparent)]
    ClientUpdateError(#[from] ClientUpdateError),
    /// Could not find welcome info for this sender and/or this epoch.
    #[error("Could not find welcome info for this sender and/or this epoch.")]
    NoWelcomeInfoFound,
    /// Error joining group.
    #[error(transparent)]
    JoinGroupError(#[from] JoinGroupError),
    /// Error joining connection group.
    #[error(transparent)]
    JoinConnectionGroupError(#[from] JoinConnectionGroupError),
    /// Error adding clients.
    #[error(transparent)]
    ClientAddtionError(#[from] ClientAdditionError),
    /// Error removing clients.
    #[error(transparent)]
    ClientRemovalError(#[from] ClientRemovalError),
    /// Error resyncing client.
    #[error(transparent)]
    ClientResyncError(#[from] ResyncClientError),
    /// Error self removing client.
    #[error(transparent)]
    ClientSelfRemovalError(#[from] ClientSelfRemovalError),
    /// Error deleting group.
    #[error(transparent)]
    GroupDeletionError(#[from] GroupDeletionError),
    /// Error performing group operation.
    #[error(transparent)]
    GroupOperationError(#[from] GroupOperationError),
    /// Deprecated parameter
    #[error("Deprecated param: {0}")]
    DeprecatedParam(&'static str),
}

/// Potential errors when joining a group.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum JoinGroupError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when joining a connection group.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum JoinConnectionGroupError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Unknown sender.
    #[error("Unknown sender.")]
    UnknownSender,
    /// Not a connection group.
    #[error("Not a connection group.")]
    NotAConnectionGroup,
    /// User auth key collision.
    #[error("User auth key collision.")]
    UserAuthKeyCollision,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when adding a user.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientAdditionError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    /// Missing queue config in client key package.
    #[error("Missing queue config in client key package.")]
    MissingQueueConfig,
    /// Incomplete Welcome message.
    #[error("Incomplete Welcome message.")]
    IncompleteWelcome,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when removing clients.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientRemovalError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when deleting a group.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum GroupDeletionError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when processing a self remove proposal.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ClientSelfRemovalError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when sending a message.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum MessageSendingError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
}

/// Potential errors when resyncing a client.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ResyncClientError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
    /// Error processing message.
    #[error("Error processing message.")]
    ProcessingError,
    #[error("Error merging commit: {0}")]
    MergeCommitError(#[from] MergeCommitError<StorageError<CborMlsAssistStorage>>),
}

/// Potential errors when validating a commit or proposal.
#[derive(Debug, Error)]
#[repr(u8)]
pub enum ValidationError {
    /// Invalid assisted message.
    #[error("Invalid assisted message.")]
    InvalidMessage,
}
