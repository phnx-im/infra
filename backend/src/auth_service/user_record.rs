// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use opaque_ke::ServerRegistration;
use phnxtypes::{crypto::OpaqueCiphersuite, identifiers::QualifiedUserName};

use crate::errors::StorageError;

#[derive(Debug, Clone)]
#[cfg_attr(test, derive(PartialEq, Eq))]
pub(super) struct UserRecord {
    user_name: QualifiedUserName,
    password_file: ServerRegistration<OpaqueCiphersuite>,
}

impl UserRecord {
    fn new(
        user_name: QualifiedUserName,
        password_file: ServerRegistration<OpaqueCiphersuite>,
    ) -> Self {
        Self {
            user_name,
            password_file,
        }
    }

    pub(super) async fn new_and_store(
        connection: impl sqlx::PgExecutor<'_>,
        user_name: &QualifiedUserName,
        opaque_record: &ServerRegistration<OpaqueCiphersuite>,
    ) -> Result<Self, StorageError> {
        let user_record = Self::new(user_name.clone(), opaque_record.clone());
        user_record.store(connection).await?;
        Ok(user_record)
    }

    #[cfg(test)]
    pub(super) fn user_name(&self) -> &QualifiedUserName {
        &self.user_name
    }

    pub(super) fn into_password_file(self) -> ServerRegistration<OpaqueCiphersuite> {
        self.password_file
    }
}

pub(crate) mod persistence {
    use phnxtypes::{
        codec::{BlobDecoded, BlobEncoded},
        identifiers::QualifiedUserName,
    };
    use sqlx::{PgExecutor, query, query_scalar};

    use crate::errors::StorageError;

    use super::UserRecord;

    impl UserRecord {
        /// Loads the AsUserRecord for a given UserName. Returns None if no AsUserRecord
        /// exists for the given UserId.
        pub(in crate::auth_service) async fn load(
            connection: impl PgExecutor<'_>,
            user_name: &QualifiedUserName,
        ) -> Result<Option<UserRecord>, StorageError> {
            let record = query_scalar!(
                r#"SELECT
                    password_file AS "password_file: _"
                FROM as_user_records
                WHERE user_name = $1"#,
                user_name.to_string(),
            )
            .fetch_optional(connection)
            .await?;
            Ok(record.map(|BlobDecoded(password_file)| {
                UserRecord::new(user_name.clone(), password_file)
            }))
        }

        /// Create a new user with the given user name. If a user with the given user
        /// name already exists, an error is returned.
        pub(super) async fn store(
            &self,
            connection: impl PgExecutor<'_>,
        ) -> Result<(), StorageError> {
            let password_file = BlobEncoded(&self.password_file);
            query!(
                "INSERT INTO as_user_records (user_name, password_file) VALUES ($1, $2)",
                self.user_name.to_string(),
                password_file as _,
            )
            .execute(connection)
            .await?;
            Ok(())
        }

        /// Deletes the AsUserRecord for a given UserId. Returns true if a AsUserRecord
        /// was deleted, false if no AsUserRecord existed for the given UserId.
        ///
        /// The storage provider must also delete the following:
        ///  - All clients of the user
        ///  - All enqueued messages for the respective clients
        ///  - All key packages for the respective clients
        pub(in crate::auth_service) async fn delete(
            connection: impl PgExecutor<'_>,
            user_name: &QualifiedUserName,
        ) -> Result<(), sqlx::Error> {
            // The database cascades the delete to the clients and their connection packages.
            query!(
                "DELETE FROM as_user_records WHERE user_name = $1",
                user_name.to_string()
            )
            .execute(connection)
            .await?;
            Ok(())
        }
    }

    #[cfg(test)]
    pub(crate) mod tests {
        use std::sync::LazyLock;

        use opaque_ke::{
            ClientRegistration, ClientRegistrationFinishParameters, ServerRegistration, ServerSetup,
        };
        use phnxtypes::{codec::PhnxCodec, crypto::OpaqueCiphersuite};
        use rand::{CryptoRng, RngCore, SeedableRng, rngs::StdRng};
        use sqlx::PgPool;
        use uuid::Uuid;

        use super::*;

        pub(crate) fn generate_password_file(
            user_name: &QualifiedUserName,
            rng: &mut (impl CryptoRng + RngCore),
        ) -> anyhow::Result<ServerRegistration<OpaqueCiphersuite>> {
            let password = b"password";

            let server_setup = ServerSetup::new(rng);
            let client_registration_start_result =
                ClientRegistration::start(rng, password).unwrap();
            let server_registration_start_result = ServerRegistration::start(
                &server_setup,
                client_registration_start_result.message,
                user_name.to_string().as_bytes(),
            )
            .unwrap();
            let client_registration_finish_result = client_registration_start_result
                .state
                .finish(
                    rng,
                    password,
                    server_registration_start_result.message,
                    ClientRegistrationFinishParameters::default(),
                )
                .unwrap();

            Ok(ServerRegistration::finish(
                client_registration_finish_result.message,
            ))
        }

        pub(crate) async fn store_random_user_record(pool: &PgPool) -> anyhow::Result<UserRecord> {
            let user_name: QualifiedUserName = format!("{}@example.com", Uuid::new_v4()).parse()?;
            let password_file = generate_password_file(&user_name, &mut rand::thread_rng())?;
            let record = UserRecord {
                user_name,
                password_file,
            };
            record.store(pool).await?;
            Ok(record)
        }

        #[sqlx::test]
        async fn load(pool: PgPool) -> anyhow::Result<()> {
            let user_record = store_random_user_record(&pool).await?;

            let loaded = UserRecord::load(&pool, &user_record.user_name)
                .await?
                .expect("missing user record");
            assert_eq!(loaded, user_record);

            Ok(())
        }

        #[sqlx::test]
        async fn delete(pool: PgPool) -> anyhow::Result<()> {
            let user_record = store_random_user_record(&pool).await?;

            let loaded = UserRecord::load(&pool, &user_record.user_name)
                .await?
                .expect("missing user record");
            assert_eq!(loaded, user_record);

            UserRecord::delete(&pool, &user_record.user_name).await?;

            let loaded = UserRecord::load(&pool, &user_record.user_name).await?;
            assert!(loaded.is_none());

            Ok(())
        }

        static PASSWORD_FILE: LazyLock<ServerRegistration<OpaqueCiphersuite>> =
            LazyLock::new(|| {
                let user_name: QualifiedUserName = "alice@example.com".parse().unwrap();
                generate_password_file(
                    &user_name,
                    &mut StdRng::seed_from_u64(0x0DDB1A5E5BAD5EEDu64),
                )
                .unwrap()
            });

        #[test]
        fn test_password_file_serde_codec() {
            insta::assert_binary_snapshot!(".cbor", PhnxCodec::to_vec(&*PASSWORD_FILE).unwrap());
        }

        #[test]
        fn test_password_file_serde_json() {
            insta::assert_json_snapshot!(&*PASSWORD_FILE);
        }
    }
}
