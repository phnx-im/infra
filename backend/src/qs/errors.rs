// SPDX-FileCopyrightText: 2023 Phoenix R&D GmbH <hello@phnx.im>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

use crate::errors::{QueueError, StorageError};

use super::network_provider::NetworkProvider;
use aircommon::crypto::errors::{DecryptionError, EncryptionError, KeyGenerationError};
use thiserror::Error;
use tracing::error;

// === DS API errors ===

/// Error fetching a message from the QS.
#[derive(Error, Debug)]
pub enum QsEnqueueError<N: NetworkProvider> {
    /// Unseal error
    #[error(transparent)]
    UnsealError(#[from] DecryptionError),
    /// An error ocurred enqueueing in a fan out queue
    #[error(transparent)]
    EnqueueError(#[from] EnqueueError),
    /// An error ocurred while sending a message to the network
    #[error("An error ocurred while sending a message to the network")]
    NetworkError(N::NetworkError),
    /// Storage provider error
    #[error("Storage provider error")]
    StorageError,
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError,
    /// Invalid response
    #[error("Invalid response")]
    InvalidResponse,
}

/// Error enqueuing a fanned-out message.
#[derive(Error, Debug)]
pub enum EnqueueError {
    /// Unrecoverable implementation error
    #[error("Library Error")]
    LibraryError, // E.g. an error while encoding a message before enqueing it.
    /// Error in the underlying storage provider
    #[error("Error in the underlying storage provider")]
    Storage(StorageError),
    /// Client not found
    #[error("Client not found")]
    ClientNotFound,
    /// Queue error
    #[error("Queue error")]
    Queue(QueueError),
}

impl From<StorageError> for EnqueueError {
    fn from(error: StorageError) -> Self {
        error!(%error, "Failed to enqueue message due to storage error");
        EnqueueError::Storage(error)
    }
}

impl From<sqlx::Error> for EnqueueError {
    fn from(error: sqlx::Error) -> Self {
        error!(%error, "Failed to enqueue message due to sqlx error");
        EnqueueError::Storage(error.into())
    }
}

impl From<QueueError> for EnqueueError {
    fn from(error: QueueError) -> Self {
        error!(%error, "Failed to enqueue message due to queue error");
        match error {
            QueueError::Storage(storage_error) => storage_error.into(),
            QueueError::SequenceNumberMismatch | QueueError::PayloadReceiverClosed => {
                Self::Queue(error)
            }
        }
    }
}

impl From<EncryptionError> for EnqueueError {
    fn from(error: EncryptionError) -> Self {
        error!(%error, "Failed to enqueue message due to encryption error");
        EnqueueError::LibraryError
    }
}

// === Internal errors ===

#[derive(Debug, Error)]
pub(super) enum GenerateAndStoreError {
    #[error("Error generating signature keypair")]
    KeyGenerationError(#[from] KeyGenerationError),
    #[error("Error storing key")]
    StorageError(#[from] StorageError),
}
