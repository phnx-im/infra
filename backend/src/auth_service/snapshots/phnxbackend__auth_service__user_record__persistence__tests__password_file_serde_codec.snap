---
source: backend/src/auth_service/user_record.rs
expression: "PhnxCodec::to_vec(&*PASSWORD_FILE).unwrap()"
extension: cbor
snapshot_kind: binary
---
